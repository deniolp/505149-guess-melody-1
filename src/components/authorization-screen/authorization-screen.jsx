import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';

import {Operation} from '../../reducer/user/user';

class AuthorizationScreen extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      name: ``,
      password: ``,
    };

    this._handleSubmit = this._handleSubmit.bind(this);
    this._handleNameInput = this._handleNameInput.bind(this);
    this._handlePasswordInput = this._handlePasswordInput.bind(this);
  }

  render() {
    const {onReplayButtonClick} = this.props;
    return <section className="login">
      <div className="login__logo"><img src="img/melody-logo.png" alt="Угадай мелодию" width="186" height="83"/></div>
      <h2 className="login__title">Вы настоящий меломан!</h2>
      <p className="login__total">За 3 минуты и 25 секунд вы набрали 12 баллов (8 быстрых), совершив 3 ошибки</p>
      <p className="login__text">Хотите сравнить свой результат с предыдущими попытками? Представтесь!</p>
      <form className="login__form" action="" onSubmit={this._handleSubmit}>
        <p className="login__field">
          <label className="login__label" htmlFor="name">Логин</label>
          <input className="login__input" type="text" name="name" id="name" onChange={this._handleNameInput} required/>
        </p>
        <p className="login__field">
          <label className="login__label" htmlFor="password">Пароль</label>
          <input className="login__input" type="text" name="password" id="password" onChange={this._handlePasswordInput} required/>
          <span className="login__error">Неверный пароль</span>
        </p>
        <button className="login__button button" type="submit">Войти</button>
      </form>
      <button className="replay" type="button" onClick={onReplayButtonClick}>Сыграть ещё раз</button>
    </section>;
  }

  _handleSubmit(evt) {
    evt.preventDefault();
    const name = this.state.name;
    const password = this.state.password;

    this.props.submitForm(name, password);
  }

  _handleNameInput(evt) {
    this.setState({
      name: evt.target.value,
    });
  }

  _handlePasswordInput(evt) {
    this.setState({
      password: evt.target.value,
    });
  }
}

AuthorizationScreen.propTypes = {
  submitForm: PropTypes.func.isRequired,
  onReplayButtonClick: PropTypes.func.isRequired,
};

const mapDispatchToProps = (dispatch) => ({
  submitForm: (email, password) => dispatch(Operation.authorizeUser(email, password)),
});

export {AuthorizationScreen};

export default connect(null, mapDispatchToProps)(AuthorizationScreen);
