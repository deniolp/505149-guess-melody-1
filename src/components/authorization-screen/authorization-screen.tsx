import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {Link, withRouter} from "react-router-dom";

import {Operation} from '../../reducer/user/user';
import {getAuthError, getUser} from '../../reducer/user/selectors';
import withFormData from '../../hocs/with-form-data/with-form-data';

class AuthorizationScreen extends PureComponent {
  constructor(props) {
    super(props);

    this._handleSubmit = this._handleSubmit.bind(this);
  }
  componentDidUpdate(prevProps) {
    if (this.props.user !== prevProps.user) {
      const {history} = this.props;
      history.push(`/win`);
    }
  }

  render() {
    const {onReplayButtonClick, mistakes, authError, onChangeNameInput, onChangePasswordInput} = this.props;

    return <section className="login">
      <div className="login__logo"><img src="img/melody-logo.png" alt="Угадай мелодию" width="186" height="83"/></div>
      <h2 className="login__title">Вы настоящий меломан!</h2>
      <p className="login__total">Вы выиграли, совершив {mistakes} ошибки(-у, -ок)</p>
      <p className="login__text">Хотите сравнить свой результат с предыдущими попытками? Представтесь!</p>
      <form className="login__form" action="" onSubmit={this._handleSubmit}>
        <p className="login__field">
          <label className="login__label" htmlFor="name">Логин</label>
          <input className="login__input" type="text" name="name" id="name" onChange={onChangeNameInput} required/>
        </p>
        <p className="login__field">
          <label className="login__label" htmlFor="password">Пароль</label>
          <input className="login__input" type="text" name="password" id="password" onChange={onChangePasswordInput} required/>
          {this._getErrorElement(authError)}
        </p>
        <button className="login__button button" type="submit">Войти</button>
      </form>
      <Link
        to="/"
        className="replay"
        onClick={onReplayButtonClick}
      >Попробовать ещё раз</Link>
    </section>;
  }

  _handleSubmit(evt) {
    evt.preventDefault();
    const {name = ``, password = ``} = this.props.formData;

    this.props.submitForm(name, password);
  }

  _getErrorElement(authError) {
    return authError ? <span className="login__error" style={{
      display: `block`,
    }}>{authError}</span> : ``;
  }
}

AuthorizationScreen.propTypes = {
  submitForm: PropTypes.func.isRequired,
  onChangePasswordInput: PropTypes.func,
  onChangeNameInput: PropTypes.func,
  onReplayButtonClick: PropTypes.func.isRequired,
  mistakes: PropTypes.number.isRequired,
  authError: PropTypes.string,
  user: PropTypes.object,
  formData: PropTypes.objectOf(PropTypes.string),
  history: PropTypes.object,
};

const mapStateToProps = (state) => ({
  authError: getAuthError(state),
  user: getUser(state),
});

const mapDispatchToProps = (dispatch) => ({
  submitForm: (email, password) => dispatch(Operation.authorizeUser(email, password)),
});

export {AuthorizationScreen};

export default withFormData(connect(mapStateToProps, mapDispatchToProps)(withRouter(AuthorizationScreen)));
