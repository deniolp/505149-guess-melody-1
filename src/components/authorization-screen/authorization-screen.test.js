import React from 'react';
import renderer from 'react-test-renderer';
import {AuthorizationScreen} from '../authorization-screen/authorization-screen';

describe(`AuthorizationScreen`, () => {
  it(`renders correctly`, () => {
    const tree = renderer.create(<AuthorizationScreen
      onReplayButtonClick = {jest.fn()}
      mistakes={2}
      authError={`Bad guy`}
      onChangeNameInput={jest.fn()}
      onChangePasswordInput={jest.fn()}
      submitForm={jest.fn()}
      formData={{
        name: `hdg@ya.ru`,
        password: `uy`,
      }}
    />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
