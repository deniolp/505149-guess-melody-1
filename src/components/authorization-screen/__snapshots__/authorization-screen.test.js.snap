// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthorizationScreen renders correctly 1`] = `
<section
  className="login"
>
  <div
    className="login__logo"
  >
    <img
      alt="Угадай мелодию"
      height="83"
      src="img/melody-logo.png"
      width="186"
    />
  </div>
  <h2
    className="login__title"
  >
    Вы настоящий меломан!
  </h2>
  <p
    className="login__total"
  >
    Вы выиграли, совершив 
    2
     ошибки(-у, -ок)
  </p>
  <p
    className="login__text"
  >
    Хотите сравнить свой результат с предыдущими попытками? Представтесь!
  </p>
  <form
    action=""
    className="login__form"
    onSubmit={[Function]}
  >
    <p
      className="login__field"
    >
      <label
        className="login__label"
        htmlFor="name"
      >
        Логин
      </label>
      <input
        className="login__input"
        id="name"
        name="name"
        onChange={[MockFunction]}
        required={true}
        type="text"
      />
    </p>
    <p
      className="login__field"
    >
      <label
        className="login__label"
        htmlFor="password"
      >
        Пароль
      </label>
      <input
        className="login__input"
        id="password"
        name="password"
        onChange={[MockFunction]}
        required={true}
        type="text"
      />
      <span
        className="login__error"
        style={
          Object {
            "display": "block",
          }
        }
      >
        Bad guy
      </span>
    </p>
    <button
      className="login__button button"
      type="submit"
    >
      Войти
    </button>
  </form>
  <button
    className="replay"
    onClick={[MockFunction]}
    type="button"
  >
    Сыграть ещё раз
  </button>
</section>
`;
